### ISSUE TITLE
Threads

### ISSUE THREAD

**Issue by nitsansoffair** [nitsansoffair/Threads#1]
package bgu.spl.a2;

import java.util.LinkedList;
import java.util.List;

/**
 * this class represents a deferred result i.e., an object that eventually will
 * be resolved to hold a result of some operation, the class allows for getting
 * the result once it is available and registering a callback that will be
 * called once the result is available.
 *
 * Note for implementors: you may add methods and synchronize any of the
 * existing methods in this class *BUT* you must be able to explain why the
 * synchronization is needed. In addition, the methods you add can only be
 * private, protected or package protected - in other words, no new public
 * methods
 *
 * @param <T> the result type
 */
public class Deferred<T> {

    //Fields
    private boolean resolved;
    private T result;
    private List<Runnable> isWaitings;

    // Pacakge protected constructor
    /*pacakge*/
    public Deferred(){
        resolved = false;
        result = null;
        isWaitings = new LinkedList<>();
    }

    /**
     *
     * @return the resolved value if such exists (i.e., if this object has been
     * {@link #resolve(java.lang.Object)}ed yet
     * @throws IllegalStateException in the case where this method is called and
     * this object is not yet resolved
     */
    public synchronized T get() {
        if(result == null)
            throw new IllegalStateException();
        return result;
    }

    /**
     *
     * @return true if this object has been resolved - i.e., if the method
     * {@link #resolve(java.lang.Object)} has been called on this object before.
     */
    public synchronized boolean isResolved() {
        return resolved;
    }

    /**
     * resolve this deferred object - from now on, any call to the method
     * {@link #get()} should return the given value
     *
     * Any callbacks that were registered to be notified when this object is
     * resolved via the {@link #whenResolved(java.lang.Runnable)} method should
     * be executed before this method returns
     *
     * @param value - the value to resolve this deferred object with
     * @throws IllegalStateException in the case where this object is already
     * resolved
     */
    public synchronized void resolve(T value) {
        if ( !resolved ) {
            this.result = value;
            resolved = true;
            while (!isWaitings.isEmpty()) {
                // go over callback functions and excute them
                Runnable tmp = isWaitings.remove(0);
                tmp.run();
            }
        }
        else{ throw new IllegalStateException(); }
    }

    /**
     * add a callback to be called when this object is resolved. if while
     * calling this method the object is already resolved - the callback should
     * be called immediately
     *
     * Note that in any case, the given callback should never get called more
     * than once, in addition, in order to avoid memory leaks - once the
     * callback got called, this object should not hold its reference any
     * longer.
     *
     * @param callback the callback to be called when the deferred object is
     * resolved
     */
    public synchronized void whenResolved(Runnable callback) {
        if (resolved){
            callback.run();
        }
        else
            isWaitings.add(callback);
    }

    // package protected function returns the List of the runnable standing by to complete this Task
    /*package*/ List<Runnable> getStandByTasks(){
        return isWaitings;
    }

}

package bgu.spl.a2;

import javafx.util.Pair;
import java.util.concurrent.BlockingDeque;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * this class represents a single work stealing processor, it is
 * {@link Runnable} so it is suitable to be executed by threads.
 *
 * Note for implementors: you may add methods and synchronize any of the
 * existing methods in this class *BUT* you must be able to explain why the
 * synchronization is needed. In addition, the methods you add can only be
 * private, protected or package protected - in other words, no new public
 * methods
 *
 */
public class Processor implements Runnable {

    private final WorkStealingThreadPool pool;
    private final int id;

    /**
     * constructor for this class
     *
     * IMPORTANT:
     * 1) this method is package protected, i.e., only classes inside
     * the same package can access it - you should *not* change it to
     * public/private/protected
     *
     * 2) you may not add other constructors to this class
     * nor you allowed to add any other parameter to this constructor - changing
     * this may cause automatic tests to fail..
     *
     * @param id - the processor id (every processor need to have its own unique
     * id inside its thread pool)
     * @param pool - the thread pool which owns this processor
     */
    /*package*/ Processor(int id, WorkStealingThreadPool pool) {
        this.id = id;
        this.pool = pool;
    }

    @Override
    /* run method always try to fetch a mission for complete her
    * if there isn`t free mission can`t be steal the processor wait*/
    public void run() {
        /* the Processor is runnable holding by thread the processor search for tasks to handle
         * the processor will wait for tasks if there isn`n any task */
        while(!Thread.currentThread().isInterrupted()){
            try{
                // get Version at start of Fetching.
                AtomicInteger version = new AtomicInteger(pool.VM.getVersion());

                // Try Fetch or Steal
                Pair<Task<?> , Boolean> Optimistic = FetchOrSteal();

                // is succeed start handle and AWAKE ALL threads when finished.
                if(Optimistic.getValue().booleanValue() && Optimistic.getKey() != null){
                    Optimistic.getKey().handle(this);
                    pool.AlarmClock();
                }
                else{// else, ONLY IF(!) the version the same as at start go sleep and wait for new Version.
                    if(version.get() == pool.VM.getVersion())
                       pool.goSleep();

                }
            }catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    private Pair<Task<?> , Boolean> FetchOrSteal() throws InterruptedException{
        /*return true if the processor got some task to excute else false */
            // first try Fetch from my Queue
        BlockingDeque myQueue = pool.getProcessorDeque(id);
        if (myQueue.size() > 0)
            return new Pair(myQueue.pollFirst(), new Boolean(true));
            // if cannot fetch from mine -> steal.
        else
        {
            boolean stole = pool.IloveSteal(id);
            if (stole)
            {
                return new Pair(myQueue.pollFirst(), new Boolean(true));
            }
            return new Pair(null, new Boolean(false));
        }
    }

    /*function add new Tasks to the Local Task queue of current processor*/
    /*package*/ void AddNewTask(Task<?>... task)
    {

        BlockingDeque myQueue = pool.getProcessorDeque(id);
        for (Task e : task)
        {
            myQueue.add(e);
        }
        pool.AlarmClock();
    }// Add New Task

}

package bgu.spl.a2;

import java.util.Collection;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * an abstract class that represents a task that may be executed using the
 * {@link WorkStealingThreadPool}
 *
 * Note for implementors: you may add methods and synchronize any of the
 * existing methods in this class *BUT* you must be able to explain why the
 * synchronization is needed. In addition, the methods you add to this class can
 * only be private!!!
 *
 * @param <R> the task result type
 */
public abstract class Task<R> {

    // fields
    private Deferred <R> Result = new Deferred<R>();
    private Processor myProcessor;
    private Runnable excuteWhenCanResolve;
    private AtomicInteger numOfDependecies;
    private boolean started = false;

    /**
     * start handling the task - note that this method is protected, a handler
     * cannot call it directly but instead must use the
     * {@link #handle(bgu.spl.a2.Processor)} method
     */
    protected abstract void start();

    /**
     *
     * start/continue handling the task
     *
     * this method should be called by a processor in order to start this task
     * or continue its execution in the case where it has been already started,
     * any sub-tasks / child-tasks of this task should be submitted to the queue
     * of the handler that handles it currently
     *
     * IMPORTANT: this method is package protected, i.e., only classes inside
     * the same package can access it - you should *not* change it to
     * public/private/protected
     *
     * @param handler the handler that wants to handle the task
     */
    /*package*/ final void handle(Processor handler) {
        /* 2 cases - 1. Task already started, then the task now ready to be executed
        *            2. otherwise start task*/
        // if Task already started, then excute.
        if(started) {
            excute();
        }

        // else start handle the start
        else{
            myProcessor = handler;
            started = true;
            start();
        }
    }

    /**
     * This method schedules a new task (a child of the current task) to the
     * same processor which currently handles this task.
     *
     * @param task the task to execute
     */
    protected final void spawn(Task<?>... task) {
            myProcessor.AddNewTask(task);
    }

    /**
     * add a callback to be executed once *all* the given tasks results are
     * resolved
     *
     * Implementors note: make sure that the callback is running only once when
     * all the given tasks completed.
     *
     * @param tasks
     * @param callback the callback to execute once all the results are resolved
     */
    protected final void whenResolved(Collection<? extends Task<?>> tasks, Runnable callback) {
        /*save the collection size from each task resolved decrement by 1
        * when we reach 0 ADD THE TASK AGAIN..*/

        // set number of dependencies of this Task
        numOfDependecies = new AtomicInteger(tasks.size());

        // add callback for each task.
        for (Task<?> e : tasks){
            e.getResult().whenResolved(() ->
                                            {// call back function
                                                /* Note -> we must synchronized this method beacuse it possible that more
                                                   than one thread tries to add the same task
                                                   */
                                                synchronized (this)
                                                {
                                                    numOfDependecies.getAndDecrement();
                                                    if (numOfDependecies.get() == 0)
                                                        myProcessor.AddNewTask(this);
                                                }
                                            });
        }
        // set callback when Task need to be excuted
        excuteWhenCanResolve = callback;
    }

    /**
     * resolve the internal result - should be called by the task derivative
     * once it is done.
     *
     * @param result - the task calculated result
     */
    protected final void complete(R result) {
       try{
        Result.resolve(result);}
        catch(Exception e){e.printStackTrace();}
    }

    /**
     * @return this task deferred result
     */
    public final Deferred<R> getResult() {
        return Result;
    }

    /* private function excute Task then can be solved
    * Note that only tasks waiting for subTasks can reach this function*/
    private void excute(){
        excuteWhenCanResolve.run();
    }
}

package bgu.spl.a2;

/**
 * Describes a monitor that supports the concept of versioning - its idea is
 * simple, the monitor has a version number which you can receive via the method
 * {@link #getVersion()} once you have a version number, you can call
 * {@link #await(int)} with this version number in order to wait until this
 * version number changes.
 *
 * you can also increment the version number by one using the {@link #inc()}
 * method.
 *
 * Note for implementors: you may add methods and synchronize any of the
 * existing methods in this class *BUT* you must be able to explain why the
 * synchronization is needed. In addition, the methods you add can only be
 * private, protected or package protected - in other words, no new public
 * methods
 */
public class VersionMonitor {

    // Fields
    private int Version;

    /*package*/ VersionMonitor (){
        Version = 0;
    }

    // returns current version
    public int getVersion() {
        return Version;
    }

    // increment by one the version
    public void inc() {
        /* Note! -> we must synchronized this function ++ operator is 3 Atomic operations
                    beacuse version is changing frequently.
         */
        synchronized (this) {
            Version++;
            notifyAll();
        }
    }

    public synchronized void await(int version) throws InterruptedException {
        while (Version == version){
           try{ wait();} catch (InterruptedException e){throw e;}
        }
    }
}

package bgu.spl.a2;

import java.util.Date;
import java.util.Random;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * represents a work stealing thread pool - to understand what this class does
 * please refer to your assignment.
 *
 * Note for implementors: you may add methods and synchronize any of the
 * existing methods in this class *BUT* you must be able to explain why the
 * synchronization is needed. In addition, the methods you add can only be
 * private, protected or package protected - in other words, no new public
 * methods
 */
public class WorkStealingThreadPool {

    /* fields */
    /* package */ VersionMonitor VM;
    private BlockingDeque<Task<?>>[] tasksOfProcessors;
    private Thread[] Processors;
    private Random rnd;
    private volatile boolean isRuning;

    /* Getters */
    /*pacakge*/ BlockingDeque<Task<?>> getProcessorDeque(int id){
        return tasksOfProcessors[id];
    }

    /**
     * creates a {@link WorkStealingThreadPool} which has nthreads
     * {@link Processor}s. Note, threads should not get started until calling to
     * the {@link #start()} method.
     *
     * Implementors note: you may not add other constructors to this class nor
     * you allowed to add any other parameter to this constructor - changing
     * this may cause automatic tests to fail..
     *
     * @param nthreads the number of threads that should be started by this
     * thread pool
     */
    public WorkStealingThreadPool(int nthreads) {
        // pool is shutdown
        isRuning = false;

        // get random seed of Random class
        rnd = new Random(new Date().getTime());

        // creating new monitor - STARTING FROM ZERO!!
        VM = new VersionMonitor();

        // queues threads and etc`
        tasksOfProcessors = new BlockingDeque[nthreads];
        Processors = new Thread [nthreads];
        CreateAll(nthreads);
    }

    /* private function Create all I need for constructor*/
    private void CreateAll(int nThreads){
        for(int i = 0 ; i < nThreads ; i++){
            Processor tmp = new Processor(i , this);
            tasksOfProcessors[i] = new LinkedBlockingDeque<>();
            Processors[i] = new Thread(tmp);
        }
    }

    /**
     * submits a task to be executed by a processor belongs to this thread pool
     *
     * @param task the task to execute
     */
    public void submit(Task<?> task) {
        if(isRuning) {
            int sumbit = rnd.nextInt(Processors.length);
            // adding at random queue
            tasksOfProcessors[sumbit].add(task);
            // notify all threads that there is task to be handle
            AlarmClock();
        }
    }

    /**
     * closes the thread pool - this method interrupts all the threads and wait
     * for them to stop - it is returns *only* when there are no live threads in
     * the queue.
     *
     * after calling this method - one should not use the queue anymore.
     *
     * @throws InterruptedException if the thread that shut down the threads is
     * interrupted
     * @throws UnsupportedOperationException if the thread that attempts to
     * shutdown the queue is itself a processor of this queue
     */
    public void shutdown() throws InterruptedException {
        /* interrupt al threads */
        for (Thread t : Processors){
            if(Thread.currentThread() == t)
                throw new UnsupportedOperationException();
        }
        isRuning = false;
        for (Thread p : Processors)
        {
            p.interrupt();
            p.join();
        }
    }

    /**
     * start the threads belongs to this thread pool
     */
    public void start() {
        /* if the pool started already do nothing */
        if (!isRuning) {
            isRuning = true;
            for (Thread Processor : Processors) {
                Processor.start();
            }
        }
    }

    /**
     * start from stealer + 1 location and trying steal from another processors.
     * this is Pcakage protected function ca be called only by objects at this
     * Package
     * @param stealer -> the id of the stealer
     * @return true if the processor stole false other wise
     */
    /* pacakge */ boolean IloveSteal( int stealer ) {
        /* function trying steal from another processor return true -> stole
        * false other wise*/
        boolean ans = false;
        for(int i = (stealer + 1) % Processors.length ; !ans && i != stealer ; i = (i + 1 ) % Processors.length ){
            if(tasksOfProcessors [i].size() > 0) {
                ans = Steal(stealer, i);
            }
        }
        return ans;
    }

    /* function stealing from victim Processor */
    private boolean Steal(int stealer, int victim) {
        boolean ans;
        int maxSteals = (tasksOfProcessors [victim].size() / 2);
        while (maxSteals != 0 && tasksOfProcessors[victim].size() > 0){
            Task stolenTask = tasksOfProcessors[victim].pollLast();
            if(stolenTask != null)
                tasksOfProcessors[stealer].add(stolenTask);
            maxSteals--;
        }
        ans = (tasksOfProcessors[stealer].size() > 0);
        return ans;
    }

    /* wake up al threads */
    /* package */ void AlarmClock(){
            VM.inc();
    }

    /* package */ void goSleep() throws InterruptedException {
        try
        {
            VM.await(VM.getVersion());
        }catch (InterruptedException e){ throw e;}
        }
    }

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package bgu.spl.a2.test;

import bgu.spl.a2.Task;
import bgu.spl.a2.WorkStealingThreadPool;
import javafx.util.Pair;

import java.util.*;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class MergeSort extends Task<int[]> {

    private final int[] array;

    public MergeSort(int[] array) {
        this.array = array;
    }

    @Override
    /* start function */
    protected void start() {
                /* 2 cases - array.length > 1 then spwan 2 new subTasks
        *            else complete Task
        *            */
        if(array.length == 1){
            complete( array );
        }
        else{
            // case - 2 create new subArrays
            List<MergeSort> tasks = new ArrayList<>();
            Pair<int[] , int[] > subArrays = GetHalfArrays();
            MergeSort leftArray = new MergeSort(subArrays.getKey());
            MergeSort rightArray = new MergeSort(subArrays.getValue());

            // add 2 new Tasks to when resolve list
            tasks.add(leftArray);
            tasks.add(rightArray);

            // call when resolved function to wait until resolved
            whenResolved( tasks, () ->{

                                        int[] Arr1 = tasks.remove(0).getResult().get();
                                        int[] Arr2 = tasks.remove(0).getResult().get();
                                        Merge(Arr1 , Arr2);
                                        complete(array);
            });

            // spwan the new tasks to the queue of current processors
            spawn(leftArray , rightArray);
        }
    }

    /* private function Merge 2 sorted arrays */
    private void Merge(int[] arr1, int[] arr2) {
        int Arr1Index = 0 , Arr22Index = 0;
        for(int i = 0 ; i < array.length ; i++)
        {
            /* 3 cases - 1. Arr2Index OutOfBounds
            *             2. Arr1Index outOfBounds
            *             3. place the smaller number*/
            // case 1
            if(Arr22Index < arr2.length &&Arr1Index < arr1.length){
                // smaller at arr1
                if(arr1[Arr1Index] < arr2[Arr22Index]){
                    array[i] = arr1[Arr1Index];
                    Arr1Index++;
                }
                // smaleer at arr2
                else{
                    array[i] = arr2[Arr22Index];
                    Arr22Index++;
                }
            }
            else{ // case 2 + 3
                if(Arr1Index < arr1.length){
                    array[i] = arr1[Arr1Index];
                    Arr1Index++;
                }
                else{
                    array[i] = arr2[Arr22Index];
                    Arr22Index++;
                }
            }
        }
        return;
    }

    /* private function creating two sub arrays O(array.length) = O(n)*/
    private Pair<int[] , int[]> GetHalfArrays(){
        // creating 2 new arrays
        int[] firstHalf = new int[array.length / 2];
        int[] secondHalf = new int[array.length - firstHalf.length];
        for(int i = 0 ; i < secondHalf.length ; i++){
            // secondHalf can be larger the firstHalf exp: array.length = 101
            // so, firstHalf = 50 , and secondHalf = 51
            if(i < firstHalf.length)
                firstHalf[i] = array[i];
            secondHalf[secondHalf.length - 1 - i] = array[array.length - 1 - i];
        }
        // return the Sub arrays
        return new Pair(firstHalf , secondHalf);
    }

    public static void main(String[] args) throws InterruptedException {
        WorkStealingThreadPool pool = new WorkStealingThreadPool(4);
        int n = 100000; //you may check on different number of elements if you like
        int[] array = new Random().ints(n).toArray();

        MergeSort task = new MergeSort(array);

        CountDownLatch l = new CountDownLatch(1);
        pool.start();
        pool.submit(task);
        task.getResult().whenResolved(() -> {
            //warning - a large print!! - you can remove this line if you wish
            System.out.println(Arrays.toString(task.getResult().get()));
            l.countDown();
        });

        l.await();
        pool.shutdown();
    }

}
